### PR TITLE
fix #1033 Remove empty windows from windowUntil/While

### LINKED ISSUE #1033: Window Operators Create Dangling Windows
Currently, when a `Flux` ends with a value that would close a window, another window is immediately opened, and closed without ever having an element provided.  For example, the two following tests fail:

```java
@Test
public void windowWhile() {
    Flux.just("ALPHA", "#", "ALPHA", "#")
        .windowWhile(s -> !"#".equals(s))
        .as(StepVerifier::create)
        .expectNextCount(2)
        .verifyComplete();
}

@Test
public void windowUntil() {
    Flux.just("ALPHA", "#", "ALPHA", "#")
        .windowUntil("#"::equals)
        .as(StepVerifier::create)
        .expectNextCount(2)
        .verifyComplete();
}
```

I believe that windows should only be opened on the first element of a new window, which would cause these two tests to pass.  This would have the added benefit of matching how the buffer operators behave.

### PR DESCRIPTION
**Pull request by simonbasle**
@smaldini @nebhale should we defer that until 3.2?
This changes the behavior of predicate-based window, deferring the offer-and-drain of windows until a) source error or b) element being pushed into the window.

### CODE REVIEW

**Comment by nebhale** [comment]
I'm not sure if you've done this (or it's done at a later date), but the marble diagram will need to be updated as well.

File: reactor-core/src/test/java/reactor/core/publisher/FluxWindowPredicateTest.java [hunk 1]
```diff
@@ -42,6 +42,40 @@
...
+	@Test
+	public void windowUntilNoEmptyWindows() {
+		Flux.just("ALPHA", "#", "BETA", "#")
+		    .windowUntil("#"::equals)
+		    .flatMap(Flux::collectList)
+		    .as(StepVerifier::create)
+		    .assertNext(w -> assertThat(w).containsExactly("ALPHA", "#"))
+		    .assertNext(w -> assertThat(w).containsExactly("BETA", "#"))
+		    .verifyComplete();
+	}
```
Comment (nebhale) [inline reactor-core/src/test/java/reactor/core/publisher/FluxWindowPredicateTest.java]: You might also want a couple more tests for intentionally empty windows:

```java
@Test
 public void windowWhileIntentionallyEmptyWindows() {
   Flux.just("ALPHA", "#", "BETA", "#", "#")
       .windowWhile(s -> !"#".equals(s))
       .flatMap(Flux::collectList)
       .as(StepVerifier::create)
       .assertNext(w -> assertThat(w).containsExactly("ALPHA"))
       .assertNext(w -> assertThat(w).containsExactly("BETA"))
       .assertNext(w -> assertThat(w).isEmpty())
       .verifyComplete();
 }

 @Test
 public void windowUntilIntentionallyEmptyWindows() {
   Flux.just("ALPHA", "#", "BETA", "#", "#")
       .windowUntil("#"::equals)
       .flatMap(Flux::collectList)
       .as(StepVerifier::create)
       .assertNext(w -> assertThat(w).containsExactly("ALPHA", "#"))
       .assertNext(w -> assertThat(w).containsExactly("BETA", "#"))
       .assertNext(w -> assertThat(w).isEmpty())
       .verifyComplete();
 }
```

**Comment by nebhale** [comment]
I guess it depends on whether you feel that this is a bug or an enhancement.  I believe that this is a bug (a well-documented bug to be sure 😄 ) and therefore, a candidate for a 3.1.x fix.  But if you believe that this is instead an enhancement, I'd understand that, and it should be 3.2.

File: reactor-core/src/test/java/reactor/core/publisher/FluxWindowPredicateTest.java (hunk 1 above)
Comment (simonbasle) [reply reactor-core/src/test/java/reactor/core/publisher/FluxWindowPredicateTest.java]: "intentionally empty windows": ok, scratch that PR, back to the drawing board

Comment (simonbasle) [reply reactor-core/src/test/java/reactor/core/publisher/FluxWindowPredicateTest.java]: so @nebhale this is absolutely 100% about not emitting an empty window on source completion? Is the following correct:

 - [ ] starting `windowUntil` with `cutBefore true` with a delimiter will produce an empty initial window, the next one will have the delimiter + whatever values
 - [ ] starting `windowWhile` with a non-matching value will produce an empty window
 - [ ] putting several delimiters in a row in a `windowWhile` will produce intermediate empty windows?
 - [ ] only case where you don't want empty windows is `windowUntil` ending with a delimiter and `windowWhile` ending with a delimiter

Comment (nebhale) [reply reactor-core/src/test/java/reactor/core/publisher/FluxWindowPredicateTest.java]: It's a good question, and I don't think I've ever thought about it in this way (you should see my face contorting as I think this over...).

I view the boundary predicate as indicating the end of a window, but _not implicitly the start of the next window_.  Only something that is explicitly window-related (i.e. new element or another boundary) can start a new window (and possibly close it immediately).

To be more concrete, I'd say that for `windowWhile()` the boundary predicate **can** implicitly open a _previous_ window, but **cannot** implicitly open a _next_ window.  For `windowUntil()` the rule holds (it can open _previous_ but not _next_) however when the **boundary is processed as an item** (either cut before or cut after) it may **also** force a new window (either _previous_ or _next_) to be opened.

Addressing your checklist, all four are `true`, which I think (hope) is consistent with what I've stated above.  In other words, I believe that the only contentious bit of the implementation in `master` is that boundaries implicitly open the _next_ window.

```java
@Test
public void windowUntilCutBefore() {
    Flux.just("#", "#", "ALPHA", "#", "#", "BETA", "#", "#")
        .windowUntil("#"::equals, true)
        .flatMap(Flux::collectList)
        .as(StepVerifier::create)
        .assertNext(w -> assertThat(w).isEmpty())
        .assertNext(w -> assertThat(w).containsExactly("#"))
        .assertNext(w -> assertThat(w).containsExactly("#", "ALPHA"))
        .assertNext(w -> assertThat(w).containsExactly("#"))
        .assertNext(w -> assertThat(w).containsExactly("#", "BETA"))
        .assertNext(w -> assertThat(w).containsExactly("#"))
        .assertNext(w -> assertThat(w).containsExactly("#"))
        .verifyComplete();
}

@Test
public void windowUntilCutAfter() {
    Flux.just("#", "#", "ALPHA", "#", "#", "BETA", "#", "#")
        .windowUntil("#"::equals, false)
        .flatMap(Flux::collectList)
        .as(StepVerifier::create)
        .assertNext(w -> assertThat(w).containsExactly("#"))
        .assertNext(w -> assertThat(w).containsExactly("#"))
        .assertNext(w -> assertThat(w).containsExactly("ALPHA", "#"))
        .assertNext(w -> assertThat(w).containsExactly("#"))
        .assertNext(w -> assertThat(w).containsExactly("BETA", "#"))
        .assertNext(w -> assertThat(w).containsExactly("#"))
        .verifyComplete();
}

@Test
public void windowWhile() {
    Flux.just("#", "#", "ALPHA", "#", "#", "BETA", "#", "#")
        .windowWhile(s -> !"#".equals(s))
        .flatMap(Flux::collectList)
        .as(StepVerifier::create)
        .assertNext(w -> assertThat(w).isEmpty())
        .assertNext(w -> assertThat(w).isEmpty())
        .assertNext(w -> assertThat(w).containsExactly("ALPHA"))
        .assertNext(w -> assertThat(w).isEmpty())
        .assertNext(w -> assertThat(w).containsExactly("BETA"))
        .assertNext(w -> assertThat(w).isEmpty())
        .verifyComplete();
}
```
